sdk/dotnet/Pulumi/Testing/IMocks.cs

```diff
@@ -0,0 +1,34 @@
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Pulumi.Testing
+{
+    /// <summary>
+    /// Hooks to mock the engine that provide test doubles for offline unit testing of stacks.
+    /// </summary>
+    public interface IMocks
+    {
+        /// <summary>
+        /// Invoked when a new resource is created by the program.
+        /// </summary>
+        /// <param name="type">Resource type name.</param>
+        /// <param name="name">Resource name.</param>
+        /// <param name="inputs">Dictionary of resource input properties.</param>
+        /// <param name="provider">Provider.</param>
+        /// <param name="id">Resource identifier.</param>
+        /// <returns>A tuple of a resource identifier and resource state. State can be either a POCO
+        /// or a dictionary bag.</returns>
+        Task<(string id, object state)> NewResourceAsync(string type, string name,
+            ImmutableDictionary<string, object> inputs, string? provider, string? id);
+
+        /// <summary>
+        /// Invoked when the program needs to call a provider to load data (e.g., to retrieve an existing
+        /// resource).
+        /// </summary>
+        /// <param name="token">Function token.</param>
+        /// <param name="args">Dictionary of input arguments.</param>
+        /// <param name="provider">Provider.</param>
+        /// <returns>Invocation result, can be either a POCO or a dictionary bag.</returns>
+        Task<object> CallAsync(string token, ImmutableDictionary<string, object> args, string? provider);
+    }
```

Comment:
can we have a Default impl of this with virtuals/delegates for all of these?  Users can then instantiate and only provide the helpers they need.  i.e. many tests will not need to provide an impl for `CallAsync`.  

If we also split out thins like ReadResource and RegisterResource, it will make it easier to only supply the functions we want.


(note: we'd still have the interface, just also have a useful concrete impl people could use).

Reply:
I think most users would use something like `Moq` to create mocks, and then they don't have to mock methods that aren't needed.
We still may need a default implementation, but I haven't spent time thinking it through yet, so it's in the Test project for now.
re `ReadResource`/`RegisterResource` -- I tried to keep it in line with JS in https://github.com/pulumi/pulumi/pull/3738

Reply:
rgd `Moq`.  That's a good point.  In that case, i would like us to do the same so people can look at our tests for how to do things.  That can come in a followup PR though.

> I tried to keep it in line with JS in #3738

Hrmm.. def eems odd to me. @pgavlin could you give some info on why you thought this was the right way to go here?

Reply:
Although... thinking about it more... i can see why it would be done this way.  This is hte high level SDK concept someone is using.  And if there's an 'id' or not lets people know if it was a read or a register... i could get behind this :)

Reply:
> could you give some info on why you thought this was the right way to go here?

Can you be more specific? I'm not sure what you're asking about.

If you're asking about the design of the mocks interface itself, I chose the `newResource` approach because in most scenarios, it doesn't matter whether or not a resource is being created or read: in either case, a user must return some sort of valid state for the resource.